### PR TITLE
8288885: Introduce a jwebserver launcher utility in test library for jtreg tests

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review for this change which adds a utility to the JDK test library to help launch the JWebServer? As noted in the JBS issue, this utility does the necessary work to make sure when the `launch()` method returns, the jwebserver is ready to receive requests. This helps remove a lot of boilerplate code from individual tests.

As part of this commit, the existing `MaxRequestTimeTest` has been migrated to use this new utility. This existing test continues to pass with this change.

tier1, tier2 and tier3 testing passed without any related issues.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288885](https://bugs.openjdk.org/browse/JDK-8288885): Introduce a jwebserver launcher utility in test library for jtreg tests

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9232/head:pull/9232` \
`$ git checkout pull/9232`

Update a local copy of the PR: \
`$ git checkout pull/9232` \
`$ git pull https://git.openjdk.org/jdk pull/9232/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9232`

View PR using the GUI difftool: \
`$ git pr show -t 9232`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9232.diff">https://git.openjdk.org/jdk/pull/9232.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
If a test needs an "in-process" HTTP server then it is easy to use the com.sun.net.httpserver API directly or use ToolProvider to find jwebserver. So I think the use-case you are looking at is JDK tests that want a HTTP server in a separate process, is that right? It's important that such tests need to be able to clean up. If I read the proposal correctly then they must call Process:destroy, is that right?

**Comment by jaikiran** [comment]
Hello Alan,

> If a test needs an "in-process" HTTP server then it is easy to use the com.sun.net.httpserver API directly or use ToolProvider to find jwebserver. So I think the use-case you are looking at is JDK tests that want a HTTP server in a separate process, is that right?

The usage of this new utility will be for those tests that test the `jwebserver` tool itself. In other words, tests that would want to run the `jwebserver` command from within the  test and then run some tests against that launched process. So you are right, this utility is for launching a separate process.
On a related note, jwebserver doesn't implement the `ToolProvider` SPI, so (for in-process usages), it cannot be found using the `ToolProvider`. My understanding, based on some previous discussions I found around this, is that it was intentional not to implement the `ToolProvider` interface for `jwebserver`.



> It's important that such tests need to be able to clean up. If I read the proposal correctly then they must call Process:destroy, is that right?

That is correct. The test which uses this utility to launch the server will be returned by a `record` which will provide the test with the access to the launched `Process`. The test is then responsible for managing that Process including destroying it. The `MaxRequestTimeTest` which uses this new utility, in this PR, has the code which does this destruction. 

Additionally, the utility itself takes the responsibility of destroying the `Process`, if the `launch()` itself failed. The `launch()` is considered as failed if after the Process has been started, the utility cannot find the server's host/port from the Process's output. In cases like that, this utility will destroy the `Process` and then throw an `IOException` from `launch()`

**Comment by jaikiran** [comment]
I've updated this PR with updates to the javadoc of this new utility to clarify that it launches a new process and the caller is responsible for destroying that process when it's no longer needed.

**Comment by AlanBateman** [comment]
test/lib/jdk/test/lib/net/SimpleHttpServer.java already exists and may have been added to support tests for MR JARs. I think it would be a bit confusing to have SimpleHttpServer and JWebServerLauncher in the same place. So maybe we should look at the existing SimpleHttpServer to see if it in the right place. My main concern with locating JWebServerLauncher in the lib tree is that I don't have a good feel yet what tests might want an out-of-process HTTP server. Multi-VM tests have historically been problematic and I would expect an in-process HTTP server to work best for most tests that need a HTTP server. No issue with refactoring MaxRequestTimeTest to split out the launching of jwebserver, it's mostly a question of whether putting this in the lib tree is the right thing or not.

**Comment by jaikiran** [comment]
> My main concern with locating JWebServerLauncher in the lib tree is that I don't have a good feel yet what tests might want an out-of-process HTTP server.

The use case that prompted me to create this utility is a very specific case where a certain change is/will be done in the `jwebserver`'s "entry point" which is the code here https://github.com/openjdk/jdk/blob/master/src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/JWebServer.java (although the javadoc of that class says "Programmatic entry point...", it actually isn't a programmatic one but is the entry point from `jwebserver` (or `jwebserver.exe`) command as setup here in the build https://github.com/openjdk/jdk/blob/master/make/modules/jdk.httpserver/Launcher.gmk). So to make sure that we really are exercising the test against that specific part of the code, I thought of writing this utility to launch that tool. Using the existing `test/lib/jdk/test/lib/net/SimpleHttpServer.java` isn't going to be of use here since it doesn't enter the `JWebServer`'s `main` method.

However, you do have a valid point. I could perhaps just invoke the `JWebServer`'s `main` method to do the same testing without requiring this utility. I suspect you are right that it's a bit too early to have this hosted in the library tree. I am inclined to shelve/withdraw this PR, but will keep it open for a few more hours just to see if there are any other comments/inputs that are against withdrawing it.

File: test/lib/jdk/test/lib/net/JWebServerLauncher.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    private static final String EXPECTED_SERVER_STARTUP_LINE = "URL ";
+
+    /**
+     * Represents a launched jwebserver
+     *
+     * @param process       The jwebserver process
+     * @param serverAddr    The address on which the server is listening
+     * @param processOutput The server's stdout and stderr output
+     */
+    public record JWebServerProcess(Process process, InetSocketAddress serverAddr,
```
Comment (dfuch) [inline test/lib/jdk/test/lib/net/JWebServerLauncher.java:55]: Maybe JWebServerProcess could be AutoCloseable (where close() would destroy the process), which would make it possible to use it in a try-with-resource block?

**Review by dfuch — CHANGES_REQUESTED** [review]

**Comment by jaikiran** [comment]
I'm withdrawing this PR, given that it's too early to make this an utility. We can revisit it if we feel the need to have such an utility to launch the https://github.com/openjdk/jdk/blob/master/src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/JWebServer.java.

**Pull request closed without merging**